### ISSUE TITLE
CSV import from LastPass failing, reason unknown

### ISSUE THREAD

**Issue by rmenessec** [keepassxreboot/keepassxc#1223]
## Expected Behavior
CSV import should import files without data loss or damage.

## Current Behavior
CSV import from CSV plaintext can silently mangle passwords. I'm still testing, but so far it fails in at least one case, and there may be others.

On my first try, I discovered that passwords will lose a character in (at least) this case: doubled double quotes. (Literal UTF-8 "" in the password text.) If this occurs in a password field, KXC will drop one of the double quotes. Other doubled characters may be dropped as well; I don't have time to perform extensive testing right now.

After trying to find workarounds, I gave up, went ahead, and imported a LastPass-exported CSV file using whatever the KXC import defaults are. I assigned column mappings manually and correctly, told it to go ahead and import, and it looks like most of my data made it. I'm going to spot-check a few entries for other doubled characters and see if I can find additional problems.

## Possible Solution
One problem I see with the current import process is that every special character option must have a character value. For example, I can't tell KXC that comments do NOT begin with a character. KXC assumes that comments _always_ begin with a character. There's no "none" option. LastPass comments don't begin with anything in exported files. There's also no "none" option for qualifying character. I would strongly suggest adding the flexibility to parse a CSV file with "none" options for qualifying characters and comments. A "smart" option for qualifying characters would be appreciated; ideally, it would attempt to Do the Right Thing in the presence of characters that are sometimes used as qualifiers, but not always.

The simplest solution for LastPass-exported data would be, obviously, to add a UI element (checkbox?) specifically to process LP CSVs as a special case. The first line of a LastPass CSV is the string:
```
url,username,password,extra,name,grouping,fav
```
The only special character appears to be the UTF-8 double quote, ". All contents past the header row appear to be literal strings or _literal strings surrounded by double quotes,_ although I haven't double-checked with a hex editor to be certain whether non-printing characters are present. Double quotes surround passwords with literal double quote characters and other non-alphanumeric characters. Simple alphanumeric passwords _do not have quotes._

Notes (the "extra" field) can contain _any literal ASCII symbols_ and may contain _multiple lines terminated by line breaks._ None of my notes contain tabs or other control characters besides line breaks, but I think there's a good chance that comment fields created from other users' data might well contain them.

At least in the case of CSV exports on UNIX, the resulting file has UNIX line endings, so the only way to tell whether you're looking at a comment field with a line break is from the context it appears in.

The last field ("fav") indicates whether the record is a favorite; that is, should appear at the top of the LP "Password Vault" when it's opened. It's not clear to me whether KXC has any similar functionality.

## Steps to Reproduce (for bugs)
Run a CSV through the import process. At least one field—unsure whether it needs to be a password field—should contain at least two contiguous UTF-8 double quotes.

## Context
Just trying to create a database by importing a CSV file exported from LastPass. I used a lot of strong, complex passwords, and a lot of my comments contain line breaks because they're perfectly legitimate in LastPass.

## Debug Info
KeePassXC - Version 2.2.2
Revision: 6d46717

Libraries:
- Qt 5.6.1
- libgcrypt 1.6.5

Operating system: Ubuntu 16.04.3 LTS
CPU architecture: x86_64
Kernel: linux 4.4.0-101-generic

Enabled extensions:
- KeePassHTTP
- Auto-Type
- YubiKey

**Comment by droidmonkey** [comment]
Some of the things you ask for in your description violate CSV format. There is no "guessing" involved in CSV. If LastPass is not exporting correct CSV then you need to clean it up before importing. A sample CSV file that fails would be very helpful in correcting this.

**Comment by rmenessec** [comment]
CSV [is not a standardized format](https://en.wikipedia.org/wiki/Comma-separated_values). There is nothing to violate. And, as I explained already, KXC is _already importing records_ near-correctly compared to _the way LastPass actually exports records._

At the moment, all I really need is a fix for the dropped double quote characters. I'm _suggesting_ a few other changes to make odd twists on CSV easier to handle.

I'm not sure why you're asking for a sample CSV file. I'm certainly not going to provide a copy of my password database. I already described the problem very precisely: if an imported password field contains two contiguous ASCII / UTF-8 double quotes, one of the double quote characters is discarded by KXC.

So, given an example input row like this:
```
http://localhost,root,""complex#)!password",,Localhost,,0
```
... KXC will destroy a double quote character in the password field, resulting in a record that looks like this:
```
http://localhost,root,"complex#)!password,,Localhost,,0
```
It doesn't seem to matter where in the field the paired double quotes are. I put them at the beginning of the field to make viewing easier.

I'm not a developer, so I'm not sure why you're putting the onus on me to sanitize LastPass' CSV output. **I can't do that.** And I certainly can't ask LastPass' parent company to do it; I'm trying to switch to KXC in the first place because their Chrome addon is bizarrely broken and their Firefox addon has been neglected to the point it's become unusable. They're not going to take time out of their schedule to make it easier for me to stop paying them for LastPass Premium.

Sanitizing the file by hand isn't feasible, either. There are 254 affected records in my password database. I have a password manager _so I don't have to_ manually update a text file when I need to change something in the password database.

Please help me.

**Comment by phoerious** [comment]
We don't ask for a sample of your actual password file, only for a dummy sample that demonstrates your problem. It's a lot easier for us to work with that than to extract all the necessary details from your lengthy description. Your two example rows already help a lot more, thanks.

**Comment by phoerious** [comment]
I must say, however, that the CSV generated by Lastpass is definitely invalid if you use quotes to surround literal field values (which may contain literally commas). It isn't specified how exactly literal quotes need to be escaped, but they need to be escaped somehow. I think Excel does that with two consecutive quotes which are reduced to a single literal quote  (in which case our behavior would be correct), others use backslashes etc. In any case, a sample that looks like this is only valid if quotes have no syntactic meaning:

    ""abc"

Otherwise, the outmost quotes indicate the literal field boundaries and are to be skipped by the parser. But if you treat "" as one literal quote, the remaining " at the end has no matching opening quote anymore. So apparently our parser tries to correct that and skips it (which is the only sensible thing it can do, otherwise everything that follows would be seen as part of the field).

A valid sample would look like this:

    """abc"

or this

    "\"abc"

or this

    ""abc

All tree would import the literal string "abc. If you want to import two quotes, you would need to repeat the escaped quote at the beginning of the field.

**Comment by hifi** [comment]
Adding enough configuration options to work around LastPass exports and adding *presets* in a combo box (one being LastPass) would help people to import their databases from different services that have csv export without knowing the exact csv quirks each of them have.

The preview is already really helpful but these kinds of things are hard to catch when most entries look fine so presets would probably help.

**Comment by seatedscribe** [comment]
I currently don't have time to verify but I'm quite sure the parser handles
double quotes coherently (which may not always be what any user expects)
because I thought a lot over this point, at that time.

Another approach could be, as a first step, adding a test among the already
devised and run it, even if I remember there is a test case which has a
double quote and doesn't get "damaged".

I would also recommend to put some effort when choosing words when
submitting issues, as this one seems to me more of a "classic" feature
request (better handling lastpass format) than an alarming silent bug :)

Cheers

**Comment by rmenessec** [comment]
@seatedscribe 
> I currently don't have time to verify but I'm quite sure the parser handles double quotes coherently

I'm not sure what "coherently" means in this case. Are you saying that reducing two double quotes to one is expected behavior?

> I would also recommend to put some effort when choosing words when submitting issues, as this one seems to me more of a "classic" feature request (better handling lastpass format) than an alarming silent bug ☺

I thought it was very _alarming_ when I exported data from LastPass as CSV plaintext and KXC imported the data without errors, warnings, or other informational messages and then I tried to log into sites via KXC browser integration and had my passwords rejected.

Given how widely used LastPass is, I just assumed that LastPass imports would have been extensively tested by now.

Since LastPass has been having outages since sometime yesterday or the day before, and I need a working solution **immediately**, I'm going to look at the requirements for moving to self-hosted [BitWarden](https://github.com/bitwarden), which appears to be fully mature.

Also, based on the extensive LastPass-related bugs filed against BitWarden—and since closed—I'd strongly suggest looking at how they handle LastPass imports. I'd also strongly suggest KXC project owners and contributors start testing against all the major services, such as 1Password, Enpass, and Dashlane. If you want to attract users, I don't think you want this to be their first experience with KXC; especially not if they're running away from LastPass' outages and bugfest, like I am.

I'm unsubscribing from this thread since I no longer have a stake in it.

**Comment by droidmonkey** [comment]
Just to be clear, yes it is appropriate to convert 2 double quotes to single quote in CSV. It is NOT appropriate to have the sequence ```""string"``` since there is an imbalance of double quotes. This is an error that you need to correct by hand before importing into ANY software.

**Comment by rmenessec** [comment]
@droidmonkey , I'm very sorry. I owe everyone in the thread an apology: the data I exported from LastPass yesterday was corrupted in some other way. (Most of?) (All of?)—the double quotes are definitely _intentional_ escape sequences generated by LastPass.

Now I _don't_ know in what exact way the import is problematic or why the more complex passwords failed to import correctly. I'm going to go back and look for more patterns and do more testing. I'm in the process of setting up some low-impact alternate password managers like pass and KeePass (the original) so I can compare results.

**Comment by rmenessec** [comment]
If anyone else has done past testing with LastPass export / KXC import, I'd appreciate any details anyone has. I'd be especially interested in knowing reasons it's failed in the past; I'd love to hear what was diagnosed and what was done to address the diagnosis.

**Comment by rmenessec** [comment]
If no one else is having problems with CSV import, I think it's safe to say that my problems have been caused by some kind of one-off flaw in LastPass. After the service came back up—possibly also after a new LastPass extension release; not sure—I was able to perform a new CSV export from the extension and I was able to get it into KXC without any significant flaws.

I _still_ don't know for certain what happened. I simply assumed that the problem must be in KXC after I fed it exports from several LastPass installs on several machines, all running different OSes, with different Firefox profiles—and also different Chrome / Chromium profiles.

On top of that, I've been keeping running CSV exports to encrypted disks from LastPass; ever since they disappeared the ability to "export" to the Firefox (or Chrome) password manager. I've never had a problem before now with reading / processing the data with any app.

Clearly, whatever happened, I should have connected it to _LastPass'_ greatly degraded software quality, rather than pinning it on a new project I've only been experimenting with, on and off, for a couple of months.

Once again, I'm _very_ sorry for wasting everyone's time. I'm also very sorry for any concern I caused with my original issue title. 😟

One final item: is there a good place to ask questions without opening issues? A Discord channel, or a mailing list; something like that? Apologies in advance if this is already documented somewhere; I didn't find anything with a quick search.
